File: 1765. Map of Highest Peak/1765. Map of Highest Peak.py
from typing import List
from collections import deque
class Solution:
    def highestPeak(self, isWater: List[List[int]]) -> List[List[int]]:
        n,m = len(isWater), len(isWater[0])
        q= deque()
        res = [[-1]*m for _ in range(n)]

        for row in range(n):
            for col in range(m):
                if isWater[row][col]==1:
                    q.append((row,col))
                    res[row][col]=0

        while q:
            row, col = q.popleft()
            hight = res[row][col]

            neighbores = [[row+1,col],[row,col+1],[row-1,col],[row,col-1]]
            for rowNei, colNei in neighbores:
                if (rowNei<0 or colNei<0 or rowNei==n or colNei==m or res[rowNei][colNei]!=-1):
                    continue
                q.append((rowNei,colNei))
                res[rowNei][colNei]=hight+1
        
        return res


sol = Solution()
print(sol.highestPeak(isWater = [[0,1],[0,0]]))
print(sol.highestPeak(isWater = [[0,0,1],[1,0,0],[0,0,0]]))

File: 1765. Map of Highest Peak/1765. Map of Highest Peak.js
/**
 * @param {number[][]} isWater
 * @return {number[][]}
 */
class CustomDeque{
    constructor(){
        this.items = []
    }
    append(element) {
        this.items.push(element)
    }
    popleft() {
        if (this.isEmpty()) {
          return "Deque is empty"
        }
        return this.items.shift()
    }
    isEmpty() {
        return this.items.length === 0
    }
    
}
var highestPeak = function(isWater) {
    const n = isWater.length
    const m = isWater[0].length
    const q = new CustomDeque()
    const res = Array.from({ length: n }, () => Array(m).fill(-1))

    for (let row=0; row<n; row++){
        for (let col=0; col<m; col++){
            if (isWater[row][col]===1){
                q.append([row,col])
                res[row][col]=0
            }
        }
    }

    while (!q.isEmpty()){
        const [row,col] = q.popleft()
        const height = res[row][col]
        const neighbores = [[row+1,col],[row,col+1],[row-1,col],[row,col-1]]
        for (const [rowNei, colNei] of neighbores) {
            if (rowNei<0 || rowNei ===n || colNei<0 || colNei===m || res[rowNei][colNei]!=-1){
                continue
            }
            q.append([rowNei,colNei])
            res[rowNei][colNei]=height+1
        }
        
    }
    return res

};
console.log(highestPeak(isWater = [[0,1],[0,0]]))
